### DISCUSSION TITLE
Downloading files from MEMFS.

### DISCUSSION THREAD

**Discussion by S10MC2015** [pyodide/pyodide#2103] [Q&A] [answered]
I have made a pyodide program that makes a epub. I want to download the epub from the emscriptem fs.

I have tried:
```py
def download(filename):
    filename = "/"+filename
    with open(filename, 'rt') as fh:
        file = fh.read()
        
    blob = Blob.new([file], {type : 'application/txt'})
    url = window.URL.createObjectURL(blob) 
    window.location.assign(url)
```
(everything is imported that is need be)

I have also tried:
```js
let txt = __pyodide_module.runPython(`                  
    with open('/book.epub', 'rt') as fh:
        txt = fh.read()
    txt
`);

const blob = new Blob([txt], {type : 'application/epub+zip'});
let url = window.URL.createObjectURL(blob);

var downloadLink = document.createElement("a");
downloadLink.href = url;
downloadLink.download = "book.epub";
document.body.appendChild(downloadLink);
downloadLink.click();
```

I keep getting this error:
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xeb in position 10: invalid continuation byte

**Comment by hoodmane** [comment, marked as answer]
The problem is that you are trying to read binary data as text. Try opening the file in binary mode: `open(filename, 'rb')`. I think this makes your code work.

But the simplest way to do this is from Javascript:
```js
new Blob([pyodide.FS.readFile(file_path)])
```
or from Python:
```py
from js import Blob
from pyodide_js import FS
Blob.new([FS.readFile(file_path)])
```
The rest of your code works as expected.

**Reply by S10MC2015** [reply]
Thank you!
